enterprise/server/remote_execution/snaploader/snaploader_test.go

```diff
@@ -78,20 +78,82 @@ func TestPackAndUnpackChunkedFiles(t *testing.T) {
-			unpacked := mustUnpack(t, ctx, loader, key, forkWorkDir, originalOpts)
+			unpacked := mustUnpack(t, ctx, loader, keys, forkWorkDir, originalOpts)
 			forkCOW := unpacked.ChunkedFiles["scratchfs"]
 			writeRandomRange(t, forkCOW)
 			forkOpts := makeFakeSnapshot(t, forkWorkDir)
 			forkOpts.ChunkedFiles = map[string]*copy_on_write.COWStore{
 				"scratchfs": forkCOW,
 			}
-			err := loader.CacheSnapshot(ctx, key, forkOpts)
+			err := loader.CacheSnapshot(ctx, keys.GetBranchKey(), forkOpts)
 			require.NoError(t, err)
 			originalOpts = forkOpts
 		}
 	}
 }
 
+func TestUnpackFallbackKey(t *testing.T) {
+	const maxFilecacheSizeBytes = 20_000_000 // 20 MB
+	ctx := context.Background()
+	env := testenv.GetTestEnv(t)
+	filecacheDir := testfs.MakeTempDir(t)
+	fc, err := filecache.NewFileCache(filecacheDir, maxFilecacheSizeBytes, false)
+	require.NoError(t, err)
+	fc.WaitForDirectoryScanToComplete()
+	env.SetFileCache(fc)
+	testcache.Setup(t, env)
+	loader, err := snaploader.New(env)
+	require.NoError(t, err)
+	workDir := testfs.MakeTempDir(t)
+
+	task := &repb.ExecutionTask{
+		Command: &repb.Command{
+			EnvironmentVariables: []*repb.Command_EnvironmentVariable{
+				// Set env vars matching a push to the main branch.
+				{Name: "GIT_BRANCH", Value: "main"},
+				{Name: "GIT_BASE_BRANCH", Value: ""},
+				{Name: "GIT_REPO_DEFAULT_BRANCH", Value: "main"},
+			},
+		},
+	}
+	keys, err := snaploader.SnapshotKeySet(task, "config-hash", "")
+	require.NoError(t, err)
+	require.Equal(t, "main", keys.GetBranchKey().GetRef())
+	require.Empty(t, keys.GetFallbackKeys())
+
+	opts := makeFakeSnapshot(t, workDir)
+	err = loader.CacheSnapshot(ctx, keys.GetBranchKey(), opts)
+	require.NoError(t, err)
+
+	forkTask := &repb.ExecutionTask{
+		Command: &repb.Command{
+			EnvironmentVariables: []*repb.Command_EnvironmentVariable{
+				// Set env vars matching a stacked PR, with the main branch as
+				// the last fallback.
+				{Name: "GIT_BRANCH", Value: "my-cool-pr"},
+				{Name: "GIT_BASE_BRANCH", Value: "my-cool-stacked-pr-base-branch"},
+				{Name: "GIT_REPO_DEFAULT_BRANCH", Value: "main"},
+			},
+		},
+	}
+	forkWorkDir := testfs.MakeDirAll(t, workDir, "VM-fallback")
+	forkKeys, err := snaploader.SnapshotKeySet(forkTask, "config-hash", "")
+	require.NoError(t, err)
+	require.Equal(t, "my-cool-pr", forkKeys.GetBranchKey().GetRef())
+	require.Len(t, forkKeys.GetFallbackKeys(), 2)
+	require.Equal(t, "my-cool-stacked-pr-base-branch", forkKeys.FallbackKeys[0].Ref)
+	require.Equal(t, "main", forkKeys.FallbackKeys[1].Ref)
+
+	// Sanity check that the branch key is not found in cache, to make sure
+	// we're actually falling back.
+	_, err = loader.GetSnapshot(ctx, &fcpb.SnapshotKeySet{BranchKey: forkKeys.GetBranchKey()})
+	require.Error(t, err)
+
+	// Now try unpacking with the complete PR key set, including fallbacks. This
+	// should succeed.
+	mustUnpack(t, ctx, loader, forkKeys, forkWorkDir, opts)
```

Comment:
Could you add a test that we don't try to update any of the fallback snapshots when we save the snapshot?

Reply:
`CacheSnapshot` only accepts a single key, not a `SnapshotKeySet` so I think it's not something that can be tested in this package.

We could add a firecracker_test case but it felt a bit overkill.